lib/std/array_list.zig

```diff
@@ -663,16 +749,95 @@ pub fn ArrayListAlignedUnmanaged(comptime T: type, comptime alignment: ?u29) typ
             self.items[n] = item;
         }
 
-        /// Insert slice `items` at index `i`. Moves `list[i .. list.len]` to
-        /// higher indicices make room.
+        /// Resize the array, adding `count` new elements at position `index`, which have `undefined` values.
+        /// The return value is a slice pointing to the newly allocated elements. The returned pointer
+        /// becomes invalid when the list is resized. Resizes list if self.capacity is not large enough.
+        pub fn addManyAtIndex(
+            self: *Self,
+            allocator: Allocator,
+            index: usize,
+            count: usize,
+        ) Allocator.Error![]T {
+            const new_len = self.items.len + count;
+            const to_move = self.items[index..];
+
+            if (self.capacity >= new_len) {
```

Comment:
Unmanaged functions with non-trivial implementations can convert to Managed and then back to Unmanaged, see `ArrayListUnmanaged.replaceRange`:

https://github.com/ziglang/zig/blob/937138cb90ae9327b0f7a932910c8d6080984f5c/lib/std/array_list.zig#L682-L686

Reply:
I have incorporated your suggestion. However, I was thinking about starting a broader conversation somewhere else. It seems to me that ArrayList could be implement by just having two fields, an UnmanagedArrayList and an Allocator, and functions could just forward the allocator. Example:

```zig
pub fn ArrayList(comptime T: type) type{
    return struct{
        unmanaged: ArrayListUnmanaged(T),
        allocator: Allocator,

        pub fn append(self: *@This(), elem: T) Allocator.Error!void{
            try self.unmanaged.append(self.allocator, elem);
        }
    };
}
```

This would deduplicate the entire code. There would be a small overhead of copying the allocator to the function call stack as a parameter, but that would probably be eliminated when the compiler inlines the call to unmanaged.append(), which should happen in ReleaseFast (and we could force inline it, if necessary). Also, in codebases that utilize both managed and unmanaged versions, the compiler could choose not to inline the calls, making the code smaller (specially in ReleaseSmall).
The only downside I see would be that we would have to access the fields indirectly through unmanaged, like: ```arrayList.unmanaged.items```. If we want to make that more ergonomic, we could just add a getter function ```items()```.